### PR TITLE
fix: Made the code compatible with default torch LR scheduler

### PR DESCRIPTION
**Pull request by TheoRudkiewicz**
For now we were using custom LR scheduler. I chose to go for classical pytorch LR schedulers. This require only calling step every epoch. However some LR scheduler could require a call to step every batch.

As a consquence I introduced new options to do step either every epoch (default value) or every batch.

### CODE REVIEW

File: src/gromo/utils/training_utils.py [hunk 1]
```diff
@@ -239,12 +239,14 @@ def gradient_descent(
...
     train_dataloader: torch.utils.data.DataLoader,
     optimizer: torch.optim.Optimizer,
-    scheduler: Any | None,
+    scheduler: torch.optim.lr_scheduler.LRScheduler | None,
     loss_function: nn.Module,
     metrics: Metric | None = None,
     batch_limit: int | None = None,
     dataloader_seed: int | None = None,
     device: torch.device = torch.device("cpu"),
+    scheduler_step_after_epoch: bool = True,
```
Comment (stephane-rivaud) [inline src/gromo/utils/training_utils.py]:
```suggestion
    scheduler_step_granulatiry: str = "epoch",
```

File: src/gromo/utils/training_utils.py [hunk 2]
```diff
@@ -307,15 +313,16 @@ def gradient_descent(
...
-        if scheduler is not None:
-            scheduler.step()
 
         # update metrics
         loss_meter.update(loss.detach(), x.size(0))
         metrics.update(y_pred.detach(), y)
 
-    if scheduler is not None and hasattr(scheduler, "epoch_step"):
-        scheduler.epoch_step()
+        if scheduler is not None and scheduler_step_after_batch:
```
Comment (stephane-rivaud) [inline src/gromo/utils/training_utils.py]:
```suggestion
        if scheduler is not None and scheduler_step_granularity == "batch":
```

File: src/gromo/utils/training_utils.py [hunk 3]
```diff
@@ -307,15 +313,16 @@ def gradient_descent(
...
+            scheduler.step()
+
+    if scheduler is not None and scheduler_step_after_epoch:
```
Comment (stephane-rivaud) [inline src/gromo/utils/training_utils.py]:
```suggestion
    if scheduler is not None and scheduler_step_granularity == "epoch"
```

File: tests/test_training_utils.py [hunk 4]
```diff
@@ -318,16 +314,31 @@ def test_with_scheduler(self):
...
+        with self.subTest("after_epoch"):
+            scheduler = _FakeScheduler()
+            gradient_descent(
+                model,
+                dl,
+                optimizer,
+                scheduler=scheduler,
+                loss_function=nn.MSELoss(reduction="mean"),
+                scheduler_step_after_batch=False,
+                scheduler_step_after_epoch=True,
```
Comment (stephane-rivaud) [inline tests/test_training_utils.py]:
```suggestion
                scheduler_step_granularity="epoch",
```

File: tests/test_training_utils.py [hunk 5]
```diff
@@ -318,16 +314,31 @@ def test_with_scheduler(self):
...
+        with self.subTest("after_batch"):
+            scheduler = _FakeScheduler()
+            gradient_descent(
+                model,
+                dl,
+                optimizer,
+                scheduler=scheduler,
+                loss_function=nn.MSELoss(reduction="mean"),
+                scheduler_step_after_batch=True,
+                scheduler_step_after_epoch=False,
```
Comment (stephane-rivaud) [inline tests/test_training_utils.py]:
```suggestion
                scheduler_step_granularity="batch",
```

File: src/gromo/utils/training_utils.py [hunk 6]
```diff
@@ -272,6 +274,10 @@ def gradient_descent(
         Default is None.
     device : torch.device, optional
         Device to use. Default is torch.device("cpu").
+    scheduler_step_after_epoch : bool, optional
+        Whether to step the scheduler after each epoch. Default is True.
+    scheduler_step_after_batch : bool, optional
+        Whether to step the scheduler after each batch. Default is False.
```
Comment (stephane-rivaud) [inline src/gromo/utils/training_utils.py]:
```suggestion
    scheduler_step_granularity : str, optional
        Whether to step the scheduler after each epoch (`"epoch"`, default) or each mini-batch (`"batch"`).
```

File: src/gromo/utils/training_utils.py [hunk 7]
```diff
@@ -239,12 +239,14 @@ def gradient_descent(
...
     optimizer: torch.optim.Optimizer,
-    scheduler: Any | None,
+    scheduler: torch.optim.lr_scheduler.LRScheduler | None,
     loss_function: nn.Module,
     metrics: Metric | None = None,
     batch_limit: int | None = None,
     dataloader_seed: int | None = None,
     device: torch.device = torch.device("cpu"),
+    scheduler_step_after_epoch: bool = True,
+    scheduler_step_after_batch: bool = False,
```
Comment (stephane-rivaud) [inline src/gromo/utils/training_utils.py]:
```suggestion
    scheduler_step_granulatiry: str = "epoch",
```

**Review by stephane-rivaud — CHANGES_REQUESTED** [review]
It should not be possible to set both options to True. Here is a workaround, unless we decide to let the user deal with this side-effect.

**Comment by TheoRudkiewicz** [comment]
> It should not be possible to set both options to True. Here is a workaround, unless we decide to let the user deal with this side-effect.

Good point, is now implemented.

**Review by stephane-rivaud — APPROVED** [review]
Accepting PR after requested changes.

**Comment by TheoRudkiewicz** [comment]
Ok I am merging without @stelladk approval as I think she only use `utils.mini_batch_gradient_descent`.
